nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+
+```ejs
+<!-- views/update.ejs -->
+<%- include("partials/errors.ejs") %>
+<form action="/users/<%= user.id %>/update" method="POST">
+  <input type="text" name="firstName" value="<%= user.firstName %>" required>
+  <input type="text" name="lastName" value="<%= user.lastName %>" required>
+  <button type="submit">Update User</button>
+</form>
+```
+
+We'll modify `index.ejs` to include a new "Update" button to make our `GET` request to the route:
+
+```ejs
+<li>
+  ID: <%= user.id %>, Name: <%= user.firstName %> <%= user.lastName %>
+  <a href="/users/<%= user.id %>/update">Update</a>
+</li>
+```
+
+Just like how we added a route for creating users, we'll add routes for updating them:
+
+```javascript
+usersRouter.get("/:id/update", usersController.usersUpdateGet);
+usersRouter.post("/:id/update", usersController.usersUpdatePost);
+```
+
+Then we'll add the logic for the requests into our controller:
+
+```javascript
+exports.usersUpdateGet = asyncHandler(async (req, res) => {
+    const user = usersStorage.getUser(req.params.id);
+    if (!user) {
+        return res.status(404).send("User not found.");
+    }
+    res.render("update", { user, errors: [] });
+});
+
+exports.usersUpdatePost = [
+    validateUser,
+    asyncHandler(async (req, res) => {
+        const user = usersStorage.getUser(req.params.id);
+        const errors = validationResult(req);
+        if (!user) {
+            return res.status(404).send("User not found.");
+        }
+        if (!errors.isEmpty()) return res.status(400).render("update", {
+            errors: errors.array(), user: user
+        });
+        const { firstName, lastName } = req.body;
+        usersStorage.updateUser(req.params.id, User(firstName, lastName));
+        res.redirect("/");
+    })
+];
+```
+
+You'll see you're now able to update users, and this looks very similar to how we create our users.
+
+Finally, let's add a way to delete users as well by starting with another form inside our `index.ejs` view:
+
+```ejs
+<!-- In views/index.ejs -->
+<ul>
+  <% if (locals.users) {%>
+    <% users.forEach(function(user) { %>
+    <li>ID: <%= user.id %>, Name: <%= user.firstName %> <%= user.lastName %></li>
+    <a href="/users/<%= user.id %>/update">Update</a>
+    <!-- This time we're sending a POST request to our route, so we need a form. -->
+    <form action="/users/<%= user.id %>/delete" method="POST" style="display:inline;">
+      <button type="submit" onclick="return confirm('Are you sure you want to delete this user?');">Delete</button>
+    </form>
+    <% }); %>
+  <% } %>
+</ul>
+```
+
+Then we add the logic to handle the request into our controller:
+
+```javascript
+// Tell the server to delete a matching user, if any. Otherwise, respond with an error.
+exports.usersDeletePost = asyncHandler(async (req, res) => {
+  const user = usersStorage.getUser(req.params.id);
+  if (!user) return res.status(404).send("User not found.");
```

Comment:
Just maintaining consistency with what users read across this lesson and other lessons
```suggestion
  if (!user) {
    return res.status(404).send("User not found.");
  }
```